File: src/telegram.zig
const std = @import("std");
const http = @import("http.zig");
const json = @import("json.zig");

pub const TgMessage = struct {
    message_id: []const u8,
    chat_id: []const u8,
    chat_type: []const u8,
    chat_title: []const u8,
    sender_id: []const u8,
    sender_name: []const u8,
    text: []const u8,
    date: i64,
    mentions_bot: bool,
    reply_to_text: ?[]const u8,
    reply_to_author: ?[]const u8,
};

pub const Telegram = struct {
    token: []const u8,
    bot_username: []const u8,
    allocator: std.mem.Allocator,
    last_update_id: i64,

    pub fn init(allocator: std.mem.Allocator, token: []const u8) Telegram {
        return Telegram{
            .token = token,
            .bot_username = "",
            .allocator = allocator,
            .last_update_id = 0,
        };
    }

    fn apiUrl(self: *Telegram, buf: []u8, method: []const u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "https://api.telegram.org/bot{s}/{s}", .{ self.token, method });
    }

    pub fn connect(self: *Telegram) !void {
        var url_buf: [512]u8 = undefined;
        const url = try self.apiUrl(&url_buf, "getMe");
        var resp = try http.get(self.allocator, url);
        defer resp.deinit();

        var parsed = try json.parse(self.allocator, resp.body);
        defer parsed.deinit();

        if (json.getObject(parsed.value, "result")) |result| {
            if (json.getString(result, "username")) |username| {
                self.bot_username = try self.allocator.dupe(u8, username);
                std.log.info("Telegram bot connected: @{s}", .{self.bot_username});
            }
        }
    }

    /// Poll for new messages. Returned TgMessages are allocated with `alloc`.
    pub fn getUpdates(self: *Telegram, alloc: std.mem.Allocator) ![]TgMessage {
        var url_buf: [512]u8 = undefined;
        var params_buf: [256]u8 = undefined;
        const params = try std.fmt.bufPrint(&params_buf, "getUpdates?timeout=30&offset={d}&allowed_updates=[\"message\"]", .{self.last_update_id + 1});
        const url = try self.apiUrl(&url_buf, params);

        var resp = try http.get(self.allocator, url);
        defer resp.deinit();

        if (resp.status != .ok) return &.{};

        var parsed = try json.parse(self.allocator, resp.body);
        defer parsed.deinit();

        const result_array = json.getArray(parsed.value, "result") orelse return &.{};
        var messages = std.ArrayList(TgMessage).init(alloc);

        for (result_array) |update| {
            const update_id = json.getInt(update, "update_id") orelse continue;
            if (update_id > self.last_update_id) {
                self.last_update_id = update_id;
            }

            const msg_obj = json.getObject(update, "message") orelse continue;
            const text = json.getString(msg_obj, "text") orelse continue;
            const chat_obj = json.getObject(msg_obj, "chat") orelse continue;
            const from_obj = json.getObject(msg_obj, "from") orelse continue;

            const chat_id_int = json.getInt(chat_obj, "id") orelse continue;
            const sender_id_int = json.getInt(from_obj, "id") orelse continue;
            const msg_id_int = json.getInt(msg_obj, "message_id") orelse continue;
            const date = json.getInt(msg_obj, "date") orelse 0;

            var chat_id_buf: [32]u8 = undefined;
            const chat_id_str = try std.fmt.bufPrint(&chat_id_buf, "{d}", .{chat_id_int});
            var sender_id_buf: [32]u8 = undefined;
            const sender_id_str = try std.fmt.bufPrint(&sender_id_buf, "{d}", .{sender_id_int});
            var msg_id_buf: [32]u8 = undefined;
            const msg_id_str = try std.fmt.bufPrint(&msg_id_buf, "{d}", .{msg_id_int});

            const chat_type = json.getString(chat_obj, "type") orelse "private";
            const chat_title = json.getString(chat_obj, "title") orelse
                json.getString(from_obj, "first_name") orelse "Unknown";
            const sender_name = json.getString(from_obj, "first_name") orelse
                json.getString(from_obj, "username") orelse "Unknown";

            // Check @bot_username mentions
            var mentions_bot = false;
            if (json.getArray(msg_obj, "entities")) |entities| {
                for (entities) |entity| {
                    if (json.getString(entity, "type")) |etype| {
                        if (std.mem.eql(u8, etype, "mention")) {
                            const offset: usize = @intCast(json.getInt(entity, "offset") orelse continue);
                            const length: usize = @intCast(json.getInt(entity, "length") orelse continue);
                            if (offset + length <= text.len and length > 1) {
                                const mention = text[offset + 1 .. offset + length];
                                if (std.ascii.eqlIgnoreCase(mention, self.bot_username)) {
                                    mentions_bot = true;
                                }
                            }
                        }
                    }
                }
            }

            // Reply context
            var reply_to_text: ?[]const u8 = null;
            var reply_to_author: ?[]const u8 = null;
            if (json.getObject(msg_obj, "reply_to_message")) |reply_msg| {
                reply_to_text = json.getString(reply_msg, "text");
                if (json.getObject(reply_msg, "from")) |reply_from| {
                    reply_to_author = json.getString(reply_from, "first_name");
                }
            }

            try messages.append(TgMessage{
                .message_id = try alloc.dupe(u8, msg_id_str),
                .chat_id = try alloc.dupe(u8, chat_id_str),
                .chat_type = try alloc.dupe(u8, chat_type),
                .chat_title = try alloc.dupe(u8, chat_title),
                .sender_id = try alloc.dupe(u8, sender_id_str),
                .sender_name = try alloc.dupe(u8, sender_name),
                .text = try alloc.dupe(u8, text),
                .date = date,
                .mentions_bot = mentions_bot,
                .reply_to_text = if (reply_to_text) |t| try alloc.dupe(u8, t) else null,
                .reply_to_author = if (reply_to_author) |a| try alloc.dupe(u8, a) else null,
            });
        }

        return messages.toOwnedSlice();
    }

    pub fn sendMessage(self: *Telegram, chat_id: []const u8, text: []const u8, reply_to: ?[]const u8) !void {
        const max_len = 4000;
        var offset: usize = 0;
        while (offset < text.len) {
            const end = @min(offset + max_len, text.len);
            const chunk = text[offset..end];
            try self.sendSingleMessage(chat_id, chunk, if (offset == 0) reply_to else null);
            offset = end;
        }
    }

    fn sendSingleMessage(self: *Telegram, chat_id: []const u8, text: []const u8, reply_to: ?[]const u8) !void {
        var url_buf: [512]u8 = undefined;
        const url = try self.apiUrl(&url_buf, "sendMessage");

        const escaped = try json.escapeString(self.allocator, text);
        defer self.allocator.free(escaped);

        var body = std.ArrayList(u8).init(self.allocator);
        defer body.deinit();

        try body.writer().print("{{\"chat_id\":{s},\"text\":\"{s}\",\"parse_mode\":\"Markdown\"", .{ chat_id, escaped });
        if (reply_to) |rid| {
            try body.writer().print(",\"reply_to_message_id\":{s}", .{rid});
        }
        try body.appendSlice("}");

        var resp = try http.postJson(self.allocator, url, body.items);
        defer resp.deinit();

        if (resp.status != .ok) {
            body.clearRetainingCapacity();
            try body.writer().print("{{\"chat_id\":{s},\"text\":\"{s}\"", .{ chat_id, escaped });
            if (reply_to) |rid| {
                try body.writer().print(",\"reply_to_message_id\":{s}", .{rid});
            }
            try body.appendSlice("}");
            var resp2 = try http.postJson(self.allocator, url, body.items);
            defer resp2.deinit();
        }
    }

    pub fn sendTyping(self: *Telegram, chat_id: []const u8) !void {
        var url_buf: [512]u8 = undefined;
        const url = try self.apiUrl(&url_buf, "sendChatAction");
        var body_buf: [256]u8 = undefined;
        const body = try std.fmt.bufPrint(&body_buf, "{{\"chat_id\":{s},\"action\":\"typing\"}}", .{chat_id});
        var resp = try http.postJson(self.allocator, url, body);
        defer resp.deinit();
    }
};

File: borg-rs/crates/borg-core/src/telegram.rs
use anyhow::{Context, Result};
use reqwest::Client;
use serde_json::Value;
use std::sync::atomic::{AtomicI64, Ordering};
use tracing::{info, warn};

/// An incoming Telegram message.
#[derive(Debug, Clone)]
pub struct TgMessage {
    pub message_id: i64,
    pub chat_id: i64,
    pub chat_type: String,
    pub chat_title: String,
    pub sender_id: i64,
    pub sender_name: String,
    pub text: String,
    pub date: i64,
    pub mentions_bot: bool,
    pub reply_to_text: Option<String>,
    pub reply_to_author: Option<String>,
}

pub struct Telegram {
    pub token: String,
    pub bot_username: String,
    client: Client,
    last_update_id: AtomicI64,
}

impl Telegram {
    pub fn new(token: impl Into<String>) -> Self {
        Self {
            token: token.into(),
            bot_username: String::new(),
            client: Client::new(),
            last_update_id: AtomicI64::new(0),
        }
    }

    fn api_url(&self, method: &str) -> String {
        format!("https://api.telegram.org/bot{}/{}", self.token, method)
    }

    /// Fetch bot info and set bot_username.
    pub async fn connect(&mut self) -> Result<()> {
        let resp: Value = self
            .client
            .get(self.api_url("getMe"))
            .send()
            .await?
            .json()
            .await?;

        if let Some(username) = resp["result"]["username"].as_str() {
            self.bot_username = username.to_string();
            info!("Telegram bot connected: @{}", username);
        }
        Ok(())
    }

    /// Long-poll for new messages with timeout=2s.
    pub async fn get_updates(&self) -> Result<Vec<TgMessage>> {
        let offset = self.last_update_id.load(Ordering::Relaxed) + 1;
        let url = format!(
            "{}?timeout=2&offset={}&allowed_updates=[\"message\"]",
            self.api_url("getUpdates"),
            offset
        );

        let resp: Value = self
            .client
            .get(&url)
            .timeout(std::time::Duration::from_secs(10))
            .send()
            .await
            .context("getUpdates request")?
            .json()
            .await
            .context("getUpdates parse")?;

        let updates = match resp["result"].as_array() {
            Some(a) => a,
            None => return Ok(vec![]),
        };

        let mut messages = Vec::new();

        for update in updates {
            let update_id = update["update_id"].as_i64().unwrap_or(0);
            if update_id > self.last_update_id.load(Ordering::Relaxed) {
                self.last_update_id.store(update_id, Ordering::Relaxed);
            }

            let msg = match update["message"].as_object() {
                Some(m) => m,
                None => continue,
            };

            let text = match msg["text"].as_str() {
                Some(t) => t.to_string(),
                None => continue,
            };

            let chat_id = msg["chat"]["id"].as_i64().unwrap_or(0);
            let chat_type = msg["chat"]["type"]
                .as_str()
                .unwrap_or("private")
                .to_string();
            let chat_title = msg["chat"]["title"]
                .as_str()
                .or_else(|| msg["chat"]["username"].as_str())
                .unwrap_or("")
                .to_string();
            let sender_id = msg["from"]["id"].as_i64().unwrap_or(0);
            let first = msg["from"]["first_name"].as_str().unwrap_or("");
            let last = msg["from"]["last_name"].as_str().unwrap_or("");
            let sender_name = if last.is_empty() {
                first.to_string()
            } else {
                format!("{} {}", first, last)
            };

            let bot_name = format!("@{}", self.bot_username).to_lowercase();
            let text_lower = text.to_lowercase();
            let mentions_bot = text_lower.contains(&bot_name)
                || msg
                    .get("entities")
                    .and_then(|e| e.as_array())
                    .map(|ents| ents.iter().any(|e| e["type"] == "mention"))
                    .unwrap_or(false);

            let reply_to_text = msg
                .get("reply_to_message")
                .and_then(|r| r["text"].as_str())
                .map(|s| s.to_string());
            let reply_to_author = msg
                .get("reply_to_message")
                .and_then(|r| r["from"]["first_name"].as_str())
                .map(|s| s.to_string());

            messages.push(TgMessage {
                message_id: msg["message_id"].as_i64().unwrap_or(0),
                chat_id,
                chat_type,
                chat_title,
                sender_id,
                sender_name,
                text,
                date: msg["date"].as_i64().unwrap_or(0),
                mentions_bot,
                reply_to_text,
                reply_to_author,
            });
        }

        Ok(messages)
    }

    /// Send a text message to a chat.
    pub async fn send_message(
        &self,
        chat_id: i64,
        text: &str,
        reply_to: Option<i64>,
    ) -> Result<()> {
        for chunk in split_text(text, 4000) {
            let mut body = serde_json::json!({
                "chat_id": chat_id,
                "text": chunk,
                "parse_mode": "Markdown",
            });
            if let Some(id) = reply_to {
                body["reply_to_message_id"] = serde_json::json!(id);
            }

            let resp: Value = self
                .client
                .post(self.api_url("sendMessage"))
                .json(&body)
                .send()
                .await?
                .json()
                .await?;

            if resp["ok"] != true {
                warn!("Telegram sendMessage failed: {:?}", resp["description"]);
            }
        }
        Ok(())
    }
}

fn split_text(text: &str, limit: usize) -> Vec<String> {
    if text.len() <= limit {
        return vec![text.to_string()];
    }
    let mut chunks = Vec::new();
    let mut remaining = text;
    while remaining.len() > limit {
        let cut = remaining[..limit].rfind('\n').unwrap_or(limit);
        chunks.push(remaining[..cut].to_string());
        remaining = remaining[cut..].trim_start_matches('\n');
    }
    if !remaining.is_empty() {
        chunks.push(remaining.to_string());
    }
    chunks
}
